File: src/main/java/io/github/robertomike/constructors/Construct.java
package io.github.robertomike.constructors;

import lombok.Getter;

import java.util.AbstractList;
import java.util.ArrayList;
import java.util.List;

@Getter
public abstract class Construct<T> extends AbstractList<T> {

    protected List<T> items = new ArrayList<>();

    @Override
    public boolean add(T t) {
        return items.add(t);
    }

    public void set(T item) {
        items.clear();
        items.add(item);
    }

    @SafeVarargs
    public final void addAll(T... item) {
        items.addAll(List.of(item));
    }

    @Override
    public T get(int index) {
        return items.get(index);
    }

    @Override
    public int size() {
        return items.size();
    }
}

File: shared/src/main/kotlin/io/github/robertomike/hefesto/constructors/Construct.kt
package io.github.robertomike.hefesto.constructors

abstract class Construct<T> : AbstractMutableList<T>() {

    protected val items: MutableList<T> = ArrayList()

    override fun add(element: T): Boolean = items.add(element)

    fun set(item: T) {
        items.clear()
        items.add(item)
    }

    fun addAll(vararg item: T) {
        items.addAll(item)
    }

    override fun get(index: Int): T = items[index]

    override val size: Int
        get() = items.size

    override fun add(index: Int, element: T) {
        items.add(index, element)
    }

    override fun removeAt(index: Int): T = items.removeAt(index)

    override fun set(index: Int, element: T): T = items.set(index, element)
}
